crates/sui-storage/src/event_store.rs

```diff
@@ -0,0 +1,97 @@
+
+use flexstr::SharedStr;
+use serde_json::Value;
+
+/// One event pulled out from the EventStore
+#[allow(unused)]
+struct StoredEvent {
+    /// UTC timestamp in milliseconds
+    timestamp: u64,
+    checkpoint_num: u64,
+    /// Only present for events pertaining to specific transactions
+    tx_digest: Option<TransactionDigest>,
+    /// The variant name from SuiEvent, eg MoveEvent, Publish, etc.
+    event_type: SharedStr,
+    /// Will be None for System events
+    move_module: Option<SharedStr>,
+    /// Individual event fields.  As much as possible these should be deconstructed and flattened,
+    /// ie `{'obj': {'fieldA': 'A', 'fieldB': 'B'}}` should really be broken down to
+    // `[('obj.fieldA', 'A'), ('obj.fieldB', 'B')]
+    fields: Vec<(SharedStr, Value)>, // Change this to something based on CBOR for binary values, or our own value types for efficiency
+}
+
+/// An EventStore supports event ingestion and flexible event querying
+/// One can think of events as logs.  They represent a log of what is happening to Sui.
+/// Thus, all different kinds of events fit on a timeline, and one should be able to query for
+/// different types of events that happen over that timeline.
+trait EventStore<EventIt>
+where
+    EventIt: Iterator<Item = StoredEvent>,
+{
+    /// Adds events to the EventStore.
+    /// Semantics: events are appended, no deduplication is done.
+    /// All events input must belong to the same checkpoint number.
+    fn add_events(
+        &self,
+        events: &[EventEnvelope],
+        checkpoint_num: u64,
+    ) -> Result<(), EventStoreError>;
+
+    /// Queries for events emitted by a given transaction, returned in order emitted
+    /// NOTE: Not all events come from transactions
+    fn events_for_transaction(&self, digest: TransactionDigest)
+        -> Result<EventIt, EventStoreError>;
+
+    /// Queries for all events of a certain EventType within a given time window.
+    /// Will return at most limit of the most recent events within the window, sorted in ascending time.
+    /// May return InvalidEventType.
+    fn events_by_type(
+        &self,
+        start_time: u64,
+        end_time: u64,
+        event_type: &str,
+        limit: usize,
+    ) -> Result<EventIt, EventStoreError>;
+
+    /// Generic event iteration bounded by time.  Return in ingestion order.
+    fn event_iterator(&self, start_time: u64, end_time: u64) -> Result<EventIt, EventStoreError>;
+
+    /// Generic event iteration bounded by checkpoint number.  Return in ingestion order.
+    /// Checkpoint numbers are inclusive on both ends.
+    fn events_by_checkpoint(
+        &self,
+        start_checkpoint: u64,
+        end_checkpoint: u64,
+    ) -> Result<EventIt, EventStoreError>;
+
+    /// Queries all Move events belonging to a certain Module ID within a given time window.
+    /// Will return at most limit of the most recent events within the window, sorted in ascending time.
+    fn events_by_module_id(
```

Comment:
`module_id` itself is not unique. 
`(PackageAddress, ModuleId)` is the one we want to use here

Reply:
module_id includes package address.

On a side note, I think a much much more frequent type of query would be events_by_event_type (here event_type as in (PackageID, Module name, event struct name)

Reply:
@longbowlu I was just talking to @damirka and it seems for many Move developers it might be sufficient to search by just package address, and not necessarily module ID.  So I wonder if these should just be separate fields for more flexibility.

Reply:
Both @lxfind's and simplified would work imo. My take on it is that as a module/dapp developer I rarely have too many events (especially duplicating ones), rarely do I publish too many modules together (although I will publish a few), and it will be more than enough to have less-specific indexes, if that helps us bunch data together.

Another thought is it is more convenient to search by package rather than by module name/ID, as if someone needs to listens to events of their packages, they know what's in there and they might need it altogether.